cursor.c

```diff
@@ -277,6 +285,18 @@ PHP_METHOD(MongoCursor, hasNext)
 	if (!php_mongo_get_more(cursor TSRMLS_CC)) {
 		RETURN_FALSE;
 	}
+
+	/* Another special case. Because we had tested for *one more* above, we do
+	 * need to check whether there are actually more results before we
+	 * conclusively can say there are more results on the cursor. We try to run
+	 * getMore here, and then check whether it had changed the "num" from the
+	 * last value. The last value of "num" is now "start", so that's why we
+	 * compare it with that. And of course, only if the cursor ID is now 0 (no
+	 * more results after this). */
+	if (cursor->cursor_id == 0 && cursor->start == cursor->num) {
+		RETURN_FALSE;
+	}
```

Comment:
Is this only triggered by a getmore returning no results? Looking at `php_mongo_get_cursor_header()`, `cursor->start` is set when processing an OP_REPLY.

When does a getmore return no results (assuming we're not dealing with a tailable cursor)? I would expect the previous OP_REPLY to report that the cursor has been closed if all results could fit in a response.

One of the first checks we do is `((cursor->limit > 0 && cursor->at >= cursor->limit) || cursor->num == 0)`, so I would expect the latter `or` conditional would catch an empy response (i.e. `cursor->num == 0`).